lib/writer.js

```diff
@@ -84,11 +85,23 @@ Writer.prototype.publish = function(topic, msg, fn){
     }
   } else {
     // wait for ready and retry
-    var self = this;
+    this.publishQueue.push([topic, msg, fn, ii]);
+  }
+};
 
-    this.once('ready', function(){
-      self.publish(topic, msg, fn)
-    });
+/**
+ * Drain publish queue after the client reconnects to nsqd.
+ *
+ * @api private
+ */
+
+Writer.prototype.drainPublishQueue = function(){
+  let data;
+
+  while (data = this.publishQueue.shift()) {
+    const [topic, msg, fn] = data;
+
+    this.publish(topic, msg, fn);
```

Comment:
Note that `publish()` can enqueue again, you could refactor to prevent it.

Reply:
I can check if `this.conns.size()` is still more than 0 before each loop, stopping the drain if the connections drop.

Reply:
I think you can just check once before the loop, `conns` can change only after the function returns.

Reply:
Given that this function is called right after `self.conns.add(conn);`, I think that check is redundant since it will always be at least 1.

Reply:
Indeed, might make sense to inline this code there since `drainPublishQueue` is not meant to be "public".